state-sync/state-sync-v2/state-sync-driver/src/bootstrapper.rs

```diff
@@ -0,0 +1,848 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    driver::DriverConfiguration,
+    error::Error,
+    storage_synchronizer::{StorageStateSummary, StorageSynchronizerInterface},
+    utils,
+};
+use data_streaming_service::{
+    data_notification::{DataNotification, DataPayload, NotificationId},
+    data_stream::DataStreamListener,
+    streaming_client::{DataStreamingClient, NotificationFeedback, StreamingServiceClient},
+};
+use diem_config::config::BootstrappingMode;
+use diem_data_client::GlobalDataSummary;
+use diem_infallible::Mutex;
+use diem_logger::*;
+use diem_types::{
+    contract_event::ContractEvent,
+    epoch_change::Verifier,
+    epoch_state::EpochState,
+    ledger_info::LedgerInfoWithSignatures,
+    transaction::{
+        default_protocol::{TransactionListWithProof, TransactionOutputListWithProof},
+        Version,
+    },
+    waypoint::Waypoint,
+};
+use event_notifications::EventSubscriptionService;
+use futures::channel::oneshot;
+use std::{collections::BTreeMap, sync::Arc};
+
+/// A simple container for verified epoch states and epoch ending ledger infos
+/// that have been fetched from the network.
+struct VerifiedEpochStates {
+    // If new epoch ending ledger infos have been fetched from the network
+    fetched_epoch_ending_ledger_infos: bool,
+
+    // The highest epoch ending version fetched thus far
+    highest_fetched_epoch_ending_version: Version,
+
+    // The latest epoch state that has been verified by the node
+    latest_epoch_state: EpochState,
+
+    // A map from versions to epoch ending ledger infos fetched from the network
+    new_epoch_ending_ledger_infos: BTreeMap<Version, LedgerInfoWithSignatures>,
+
+    // If the node has successfully verified the waypoint
+    verified_waypoint: bool,
+}
+
+impl VerifiedEpochStates {
+    pub fn new(latest_epoch_state: EpochState) -> Self {
+        Self {
+            fetched_epoch_ending_ledger_infos: false,
+            highest_fetched_epoch_ending_version: 0,
+            latest_epoch_state,
+            new_epoch_ending_ledger_infos: BTreeMap::new(),
+            verified_waypoint: false,
+        }
+    }
+
+    /// Returns true iff the node has already fetched any new epoch
+    /// ending ledger infos from the network.
+    pub fn fetched_epoch_ending_ledger_infos(&self) -> bool {
+        self.fetched_epoch_ending_ledger_infos
+    }
+
+    /// Sets `fetched_epoch_ending_ledger_infos` to true
+    pub fn set_fetched_epoch_ending_ledger_infos(&mut self) {
+        self.fetched_epoch_ending_ledger_infos = true;
+    }
+
+    /// Returns true iff the node has verified the waypoint
+    pub fn verified_waypoint(&self) -> bool {
+        self.verified_waypoint
+    }
+
+    /// Sets `verified_waypoint` to true
+    pub fn set_verified_waypoint(&mut self) {
+        self.verified_waypoint = true;
+    }
```

Comment:
Would this not be set by `verify_waypoint`?  Seems strange to control something like this from the outside.  If it's for testing we should label it appropriately.

Reply:
Yeah, annoyingly it needs to be accessible/set by two different places: `verify_waypoint` and `verify_waypoint_is_satisfiable()`. The satisfiability check occurs in the bootstrapper and basically sets if it storage has already verified the waypoint in the past (e.g., we just crashed and rebooted), or if the waypoint is genesis.